From cd8aa201b6ffb3a3ae138707e0720d619982d569 Mon Sep 17 00:00:00 2001
From: "Niels Westphal (crashniels)" <nielswest@live.de>
Date: Thu, 10 Jul 2025 17:58:42 +0200
Subject: [PATCH] Clean up Gladius registers
MIME-Version: 1.0
Content-Type: multipart/mixed; boundary="------------false"

This is a multi-part message in MIME format.
--------------false
Content-Type: text/plain; charset=UTF-8; format=fixed
Content-Transfer-Encoding: 8bit


CB_HW_CONTROL change in liverpool untested. Hawaii also does this and Bonaire only the second CB_HW_CONTROL.

Mostly a copy of Liverpool (which is a copy of Bonaire) with some extras related to power management and other stuff I don't quite understand.
---
 drivers/gpu/drm/amd/amdgpu/cik.c | 202 ++++++++++++++++++++++++++-----
 1 file changed, 170 insertions(+), 32 deletions(-)


--------------false
Content-Type: text/x-patch; name="0028-Clean-up-Gladius-registers.patch"
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="0028-Clean-up-Gladius-registers.patch"

diff --git a/drivers/gpu/drm/amd/amdgpu/cik.c b/drivers/gpu/drm/amd/amdgpu/cik.c
index 98613b21fb645bff08586c47e2d28c8799b53565..7c1161ab7d1eb65c5f49c81f13504e9e5873fa4a 100644
--- a/drivers/gpu/drm/amd/amdgpu/cik.c
+++ b/drivers/gpu/drm/amd/amdgpu/cik.c
@@ -811,7 +811,7 @@ static const u32 liverpool_golden_registers[] =
 	0xf004, 0xffff0000, 0xff000000, /* CGTS_USER_TCC_DISABLE */
 	0xf080, 0xfdfc0fff, 0x00000100, /* CGTT_SPI_CLK_CTRL */
 	0x1bb6, 0x00010000, 0x00010000, /* CRTC_DOUBLE_BUFFER_CONTROL */
-	0x2684, 0x00210000, 0x00018208, /* CB_HW_CONTROL */
+	0x2684, 0x00210000, 0x00058208, /* CB_HW_CONTROL */
 	0x260d, 0xf00fffff, 0x00004400, /* DB_DEBUG2 */
 	0x16ec, 0x000000f0, 0x00000070, /* FBC_DEBUG_COMP */
 	0x263e, 0x73773777, 0x12011003, /* GB_ADDR_CONFIG */
@@ -829,7 +829,7 @@ static const u32 liverpool_golden_registers[] =
 	0x31e6, 0xffffffff, 0x00ffffbf, /* SPI_RESOURCE_RESERVE_EN_CU_0 */
 	0x31e7, 0xffffffff, 0x00ffffaf, /* SPI_RESOURCE_RESERVE_EN_CU_1 */
 	0x31e8, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_2 */
-	0x31e9, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_3*/
+	0x31e9, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_3 */
 	0x31ea, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_4 */
 	0x31eb, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_5 */
 	0x31ec, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_6 */
@@ -932,38 +932,174 @@ static const u32 liverpool_mgcg_cgcg_init[] =
 	0x1579, 0xff607fff, 0xfc000100, /* CGTT_DRM_CLK_CTRL0 */
 };
 
+static const u32 gladius_golden_spm_registers[] =
+{
+	0xc200, 0xe0ffffff, 0xe0000000 	/* mmGRBM_GFX_INDEX */
+};
+
 static const u32 gladius_golden_common_registers[] =
 {
-	mmATC_MISC_CG, 0x000c0fc0, 0x000c0200,
-	mmCB_HW_CONTROL, 0x0001f3cf, 0x00007208,	// 0x2684
-	mmCB_HW_CONTROL_2, 0x0f000000, 0x0f000000,
-	mmCB_HW_CONTROL_3, 0x000001ff, 0x00000040,
-	mmDB_DEBUG2, 0xf00fffff, 0x00000400,
-	mmPA_SC_ENHANCE, 0xffffffff, 0x20000001,
-	mmPA_SC_LINE_STIPPLE_STATE, 0x0000ff0f, 0x00000000,
-	mmPA_SC_RASTER_CONFIG, 0x3f3fffff, 0x2a00161a,
-	mmPA_SC_RASTER_CONFIG_1, 0x0000003f, 0x0000002e,
-	mmRLC_CGCG_CGLS_CTRL, 0x00000003, 0x0020003c, //0x0001003c,
-	0xec9d, 0xffffffff, 0x0001003c,//mmRLC_CGCG_CGLS_CTRL_3D, 0xffffffff, 0x0001003c,
-	mmSQ_CONFIG, 0x07f80000, 0x07180000,
-	mmTA_CNTL_AUX, 0x000f000f, 0x000b0000,
-	mmTCC_CTRL, 0x00100000, 0xf31fff7f,
-	mmTCP_ADDR_CONFIG, 0x000003ff, 0x000000f7,
-	mmTCP_CHAN_STEER_HI, 0xffffffff, 0x00000000,
-	mmVGT_RESET_DEBUG, 0x00000004, 0x00000004,
+	0xc200, 0xffffffff, 0xe0000000, /* mmGRBM_GFX_INDEX */
+	0xa0d4, 0xffffffff, 0x2a00161a, /* PA_SC_RASTER_CONFIG */
+	0xa0d5, 0xffffffff, 0x0000002e, /* PA_SC_RASTER_CONFIG_1 */
+	0x2684, 0xffffffff, 0x00018208, /* mmCB_HW_CONTROL */
+	0x263e, 0xffffffff, 0x12011003, /* mmGB_ADDR_CONFIG */
 };
 
 static const u32 gladius_golden_registers[] =
+{	
+	0xf000, 0xffff1fff, 0x96940200, /* CGTS_SM_CTRL_REG */
+	0xf003, 0xffff0001, 0xff000000, /* CGTS_TCC_DISABLE */
+	0xf004, 0xffff0000, 0xff000000, /* CGTS_USER_TCC_DISABLE */
+	0xf080, 0xfdfc0fff, 0x00000100, /* CGTT_SPI_CLK_CTRL */
+	0x1bb6, 0x00010000, 0x00010000, /* CRTC_DOUBLE_BUFFER_CONTROL */
+	0x2684, 0x00210000, 0x00058208, /* CB_HW_CONTROL */
+	0x260d, 0xf00fffff, 0x00004400, /* DB_DEBUG2 */
+	0x16ec, 0x000000f0, 0x00000070, /* FBC_DEBUG_COMP */
+	0x263e, 0x73773777, 0x12011003, /* GB_ADDR_CONFIG */
+	0xbd2, 0x73773777, 0x12010001, 	/* HDP_ADDR_CONFIG */
+	0x2285, 0xf000003f, 0x00000007, /* PA_CL_ENHANCE */
+	0x22fc, 0x00000001, 0x00000001, /* PA_SC_ENHANCE */
+	0x22c9, 0xffffffff, 0x00ffffff, /* PA_SC_FORCE_EOV_MAX_CNTS */
+	0xc281, 0x0000ff0f, 0x00000000, /* PA_SC_LINE_STIPPLE_STATE */
+	0xa293, 0x07ffffff, 0x06000000, /* PA_SC_MODE_CNTL_1 */
+	0x136, 0x00000fff, 0x00000100, 	/* SCLK_CGTT_BLK_CTRL_REG */
+	0xf9e, 0x00000001, 0x00000002, 	/* SEM_CHICKEN_BITS */
+	0x31da, 0x00000008, 0x00000008, /* SPI_RESET_DEBUG */
+	0x31dc, 0xffffffff, 0x00000800, /* SPI_RESOURCE_RESERVE_CU_0 */
+	0x31dd, 0xffffffff, 0x00000800, /* SPI_RESOURCE_RESERVE_CU_1 */
+	0x31e6, 0xffffffff, 0x00ffffbf, /* SPI_RESOURCE_RESERVE_EN_CU_0 */
+	0x31e7, 0xffffffff, 0x00ffffaf, /* SPI_RESOURCE_RESERVE_EN_CU_1 */
+	0x31e8, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_2 */
+	0x31e9, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_3 */
+	0x31ea, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_4 */
+	0x31eb, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_5 */
+	0x31ec, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_6 */
+	0x31ed, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_7 */
+	0x31ee, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_8 */
+	0x31ef, 0xffffffff, 0x00fffffe, /* SPI_RESOURCE_RESERVE_EN_CU_9 */
+	0x2300, 0x000000ff, 0x00000001, /* SQ_CONFIG */
+	0x2542, 0x00010000, 0x00010000, /* TA_CNTL_AUX */
+	0x2b03, 0xffffffff, 0x76325410, /* TCP_CHAN_STEER_LO */
+	0x535, 0xffffffff, 0x00000000, 	/* VM_CONTEXTS_DISABLE */
+};
+
+static const u32 gladius_mgcg_cgcg_init[] =
 {
-	mmGRBM_GFX_INDEX, 0xffffffff, 0xe0000000,
-	mmPA_SC_RASTER_CONFIG, 0xffffffff, 0x2a00161a,
-	mmPA_SC_RASTER_CONFIG_1, 0xffffffff, 0x0000002e,
-	mmGB_ADDR_CONFIG, 0xffffffff, 0x12011003,
-	mmSPI_RESOURCE_RESERVE_CU_0, 0xffffffff, 0x00000800,
-	mmSPI_RESOURCE_RESERVE_CU_1, 0xffffffff, 0x00000800,
-	mmSPI_RESOURCE_RESERVE_EN_CU_0, 0xffffffff, 0x00FF7FBF,
-	mmSPI_RESOURCE_RESERVE_EN_CU_1, 0xffffffff, 0x00FF7FAF,
-	0x535, 0xffffffff, 0x00000000, /* VM_CONTEXTS_DISABLE */
+	0x3108, 0xffffffff, 0xfffffffc, /* RLC_CGTT_MGCG_OVERRIDE */
+	0xc200, 0xffffffff, 0xe0000000, /* GRBM_GFX_INDEX */
+	0xf0a8, 0xffffffff, 0x00000100, /* CB_CGTT_SCLK_CTRL */
+	0xf082, 0xffffffff, 0x00000100, /* CGTT_BCI_CLK_CTRL */
+	0xf0b0, 0xffffffff, 0x00000100, /* CGTT_CP_CLK_CTRL */
+	0xf0b2, 0xffffffff, 0x00000100, /* CGTT_CPC_CLK_CTRL */
+	0xf0b1, 0xffffffff, 0x00000100, /* CGTT_CPF_CLK_CTRL */
+	0x1579, 0xffffffff, 0x00600100, /* CGTT_DRM_CLK_CTRL0 */
+	0xf0a0, 0xffffffff, 0x00000100, /* CGTT_GDS_CLK_CTRL */
+	0xf085, 0xffffffff, 0x06000100, /* CGTT_IA_CLK_CTRL */
+	0xf088, 0xffffffff, 0x00000100, /* CGTT_PA_CLK_CTRL */
+	0xf086, 0xffffffff, 0x06000100, /* CGTT_WD_CLK_CTRL */
+	0xf081, 0xffffffff, 0x00000100, /* CGTT_PC_CLK_CTRL */
+	0xf0b8, 0xffffffff, 0x00000100, /* CGTT_RLC_CLK_CTRL */
+	0xf089, 0xffffffff, 0x00000100, /* CGTT_SC_CLK_CTRL */
+	0xf080, 0xffffffff, 0x00000100, /* CGTT_SPI_CLK_CTRL */
+	0xf08c, 0xffffffff, 0x00000100, /* CGTT_SQ_CLK_CTRL */
+	0xf08d, 0xffffffff, 0x00000100, /* CGTT_SQG_CLK_CTRL */
+	0xf094, 0xffffffff, 0x00000100, /* CGTT_SX_CLK_CTRL0 */
+	0xf095, 0xffffffff, 0x00000100, /* CGTT_SX_CLK_CTRL1 */
+	0xf096, 0xffffffff, 0x00000100, /* CGTT_SX_CLK_CTRL2 */
+	0xf097, 0xffffffff, 0x00000100, /* CGTT_SX_CLK_CTRL3 */
+	0xf098, 0xffffffff, 0x00000100, /* CGTT_SX_CLK_CTRL4 */
+	0xf09f, 0xffffffff, 0x00000100, /* CGTT_TCI_CLK_CTRL */
+	0xf09e, 0xffffffff, 0x00000100, /* CGTT_TCP_CLK_CTRL */
+	0xf084, 0xffffffff, 0x06000100, /* CGTT_VGT_CLK_CTRL */
+	0xf0a4, 0xffffffff, 0x00000100, /* DB_CGTT_CLK_CTRL_0 */
+	0xf09d, 0xffffffff, 0x00000100, /* TA_CGTT_CTRL */
+	0xf0ad, 0xffffffff, 0x00000100, /* TCA_CGTT_SCLK_CTRL */
+	0xf0ac, 0xffffffff, 0x00000100, /* TCC_CGTT_SCLK_CTRL */
+	0xf09c, 0xffffffff, 0x00000100, /* TD_CGTT_CTRL */
+	0xc200, 0xffffffff, 0xe0000000, /* GRBM_GFX_INDEX */
+	0xf008, 0xffffffff, 0x00010000, /* CGTS_CU0_SP0_CTRL_REG */
+	0xf009, 0xffffffff, 0x00030002, /* CGTS_CU0_LDS_SQ_CTRL_REG */
+	0xf00a, 0xffffffff, 0x00040007, /* CGTS_CU0_TA_SQC_CTRL_REG */
+	0xf00b, 0xffffffff, 0x00060005, /* CGTS_CU0_SP1_CTRL_REG */
+	0xf00c, 0xffffffff, 0x00090008, /* CGTS_CU0_TD_TCP_CTRL_REG */
+	0xf00d, 0xffffffff, 0x00010000, /* CGTS_CU1_SP0_CTRL_REG */
+	0xf00e, 0xffffffff, 0x00030002, /* CGTS_CU1_LDS_SQ_CTRL_REG */
+	0xf00f, 0xffffffff, 0x00040007, /* CGTS_CU1_TA_CTRL_REG */
+	0xf010, 0xffffffff, 0x00060005, /* CGTS_CU1_SP1_CTRL_REG */
+	0xf011, 0xffffffff, 0x00090008, /* CGTS_CU1_TD_TCP_CTRL_REG */
+	0xf012, 0xffffffff, 0x00010000, /* CGTS_CU2_SP0_CTRL_REG */
+	0xf013, 0xffffffff, 0x00030002, /* CGTS_CU2_LDS_SQ_CTRL_REG */
+	0xf014, 0xffffffff, 0x00040007, /* CGTS_CU2_TA_CTRL_REG */
+	0xf015, 0xffffffff, 0x00060005, /* CGTS_CU2_SP1_CTRL_REG */
+	0xf016, 0xffffffff, 0x00090008, /* CGTS_CU2_TD_TCP_CTRL_REG */
+	0xf017, 0xffffffff, 0x00010000, /* CGTS_CU3_SP0_CTRL_REG */
+	0xf018, 0xffffffff, 0x00030002, /* CGTS_CU3_LDS_SQ_CTRL_REG */
+	0xf019, 0xffffffff, 0x00040007, /* CGTS_CU3_TA_SQC_CTRL_REG */
+	0xf01a, 0xffffffff, 0x00060005, /* CGTS_CU3_SP1_CTRL_REG */
+	0xf01b, 0xffffffff, 0x00090008, /* CGTS_CU3_TD_TCP_CTRL_REG */
+	0xf01c, 0xffffffff, 0x00010000, /* CGTS_CU4_SP0_CTRL_REG */
+	0xf01d, 0xffffffff, 0x00030002, /* CGTS_CU4_LDS_SQ_CTRL_REG */
+	0xf01e, 0xffffffff, 0x00040007, /* CGTS_CU4_TA_CTRL_REG */
+	0xf01f, 0xffffffff, 0x00060005, /* CGTS_CU4_SP1_CTRL_REG */
+	0xf020, 0xffffffff, 0x00090008, /* CGTS_CU4_TD_TCP_CTRL_REG */
+	0xf021, 0xffffffff, 0x00010000, /* CGTS_CU5_SP0_CTRL_REG */
+	0xf022, 0xffffffff, 0x00030002, /* CGTS_CU5_LDS_SQ_CTRL_REG */
+	0xf023, 0xffffffff, 0x00040007, /* CGTS_CU5_TA_CTRL_REG */
+	0xf024, 0xffffffff, 0x00060005, /* CGTS_CU5_SP1_CTRL_REG */
+	0xf025, 0xffffffff, 0x00090008, /* CGTS_CU5_TD_TCP_CTRL_REG */
+	0xf026, 0xffffffff, 0x00010000, /* CGTS_CU6_SP0_CTRL_REG */
+	0xf027, 0xffffffff, 0x00030002, /* CGTS_CU6_LDS_SQ_CTRL_REG */
+	0xf028, 0xffffffff, 0x00040007, /* CGTS_CU6_TA_SQC_CTRL_REG */
+	0xf029, 0xffffffff, 0x00060005, /* CGTS_CU6_SP1_CTRL_REG */
+	0xf02a, 0xffffffff, 0x00090008, /* CGTS_CU6_TD_TCP_CTRL_REG */
+	0xf02b, 0xffffffff, 0x00010000, /* CGTS_CU7_SP0_CTRL_REG */
+	0xf02c, 0xffffffff, 0x00030002, /* CGTS_CU7_LDS_SQ_CTRL_REG */
+	0xf02d, 0xffffffff, 0x00040007, /* CGTS_CU7_TA_SQC_CTRL_REG */
+	0xf02e, 0xffffffff, 0x00060005, /* CGTS_CU7_SP1_CTRL_REG */
+	0xf02f, 0xffffffff, 0x00090008, /* CGTS_CU7_TD_TCP_CTRL_REG */
+	0xf030, 0xffffffff, 0x00010000, /* CGTS_CU8_SP0_CTRL_REG */
+	0xf031, 0xffffffff, 0x00030002, /* CGTS_CU8_LDS_SQ_CTRL_REG */
+	0xf032, 0xffffffff, 0x00040007, /* CGTS_CU8_TA_CTRL_REG */
+	0xf033, 0xffffffff, 0x00060005, /* CGTS_CU8_SP1_CTRL_REG */
+	0xf034, 0xffffffff, 0x00090008, /* CGTS_CU8_TD_TCP_CTRL_REG */
+	0xf035, 0xffffffff, 0x00010000, /* CGTS_CU9_SP0_CTRL_REG */
+	0xf036, 0xffffffff, 0x00030002, /* CGTS_CU9_LDS_SQ_CTRL_REG */
+	0xf037, 0xffffffff, 0x00040007, /* CGTS_CU9_TA_CTRL_REG */
+	0xf038, 0xffffffff, 0x00060005, /* CGTS_CU9_SP1_CTRL_REG */
+	0xf039, 0xffffffff, 0x00090008, /* CGTS_CU9_TD_TCP_CTRL_REG */
+	0xf000, 0xffffffff, 0x96940200, /* CGTS_SM_CTRL_REG */
+	0x21c2, 0xffffffff, 0x00900100, /* CP_RB_WPTR_POLL_CNTL */
+	0x3109, 0xffffffff, 0x0020003f, /* RLC_CGCG_CGLS_CTRL */
+	0x1579, 0xff607fff, 0xfc000100, /* CGTT_DRM_CLK_CTRL0 */
+	/* Power Management + extras */
+	0x1401, 0x00002000, 0x00002000, /* GARLIC_FLUSH_CNTL */
+	0x3114, 0xffffffff, 0x00000032, /* mmRLC_MAX_PG_CU */ //maybe 0x00000024
+	0x2240, 0xffffffff, 0x0c020001, /* mmCC_GC_PRIM_CONFIG */
+	0x219f, 0xffffffff, 0x00008000, /* mmCP_BUSY_STAT */
+	0x2088, 0xffffffff, 0x00000000, /* mmCP_CPF_BUSY_STAT */
+	0x2089, 0xffffffff, 0x00000000, /* mmCP_CPF_STALLED_STAT1 */
+	0x2087, 0xffffffff, 0x00000000, /* mmCP_CPF_STATUS */
+	0x25c1, 0xffffffff, 0x00000000, /* mmGDS_CNTL_STATUS */
+	0x25c3, 0xffffffff, 0x00000000, /* mmGDS_PROTECTION_FAULT */
+	0x2002, 0xffffffff, 0x00000000, /* mmGRBM_STATUS2 */
+	0x2003, 0xffffffff, 0x00000010, /* mmGRBM_PWR_CNTL */
+	0x2004, 0xffffffff, 0x00000000, /* mmGRBM_STATUS */
+	0x2005, 0xffffffff, 0x00000000, /* mmGRBM_STATUS_SE0 */
+	0x2006, 0xffffffff, 0x00000000, /* mmGRBM_STATUS_SE1 */
+	0x200e, 0xffffffff, 0x00000000, /* mmGRBM_STATUS_SE2 */
+	0x200f, 0xffffffff, 0x00000000, /* mmGRBM_STATUS_SE3 */
+	0x219e, 0xffffffff, 0x00000000, /* mmCP_STALLED_STAT2 */
+	0x219c, 0xffffffff, 0x00000000, /* mmCP_STALLED_STAT3 */
+	0x21a0, 0xffffffff, 0x00000000, /* mmCP_STAT */
+	0xc2a9, 0xffffffff, 0x00000020, /* mmPA_SC_HP3D_TRAP_SCREEN_H */
+	0xc2aa, 0xffffffff, 0x00000080, /* mmPA_SC_HP3D_TRAP_SCREEN_V */
+	0xc2a1, 0xffffffff, 0x00000001, /* mmPA_SC_P3D_TRAP_SCREEN_H */
+	0xc2a2, 0xffffffff, 0x00000000, /* mmPA_SC_P3D_TRAP_SCREEN_V */
+	0xc2b1, 0xffffffff, 0x00000001, /* mmPA_SC_TRAP_SCREEN_H */
+	0xc2b2, 0xffffffff, 0x00000000, /* mmPA_SC_TRAP_SCREEN_V */
+	0x24d7, 0xffffffff, 0x0003ffff, /* mmSPI_PG_ENABLE_STATIC_CU_MASK */
+	0xf09b, 0xffffffff, 0x00000000, //DONE maybe 0x00000100
 };
 
 static const u32 godavari_golden_registers[] =
@@ -1093,16 +1229,18 @@ static void cik_init_golden_registers(struct amdgpu_device *adev)
 						 ARRAY_SIZE(liverpool_golden_spm_registers));
 		break;
 	case CHIP_GLADIUS:
+		amdgpu_device_program_register_sequence(adev,
+						 gladius_mgcg_cgcg_init,
+						 ARRAY_SIZE(gladius_mgcg_cgcg_init));
 		amdgpu_device_program_register_sequence(adev,
 						 gladius_golden_registers,
 						 ARRAY_SIZE(gladius_golden_registers));
 		amdgpu_device_program_register_sequence(adev,
 						 gladius_golden_common_registers,
 						 ARRAY_SIZE(gladius_golden_common_registers));
-		// TODO (ps4patches): Why is this using the Hawaii golden spm registers, that seems wrong?
 		amdgpu_device_program_register_sequence(adev,
-						 hawaii_golden_spm_registers,
-						 ARRAY_SIZE(hawaii_golden_spm_registers));
+						 gladius_golden_spm_registers,
+						 ARRAY_SIZE(gladius_golden_spm_registers));
 		break;
 	default:
 		break;
@@ -2568,4 +2706,4 @@ int cik_set_ip_blocks(struct amdgpu_device *adev)
 		return -EINVAL;
 	}
 	return 0;
-}
+}
\ No newline at end of file

--------------false--


